cpu/qn908x/periph/uart.c

```diff
@@ -0,0 +1,293 @@
+     *   = |remainder(T + A * B / 2, A * B) - A * B / 2|
+     */
+    for (uint8_t osrval_p1 = UART_OSRVAL_MAX + 1; osrval_p1 > 8; osrval_p1--) {
+        /* Initial value of (OSRVAL + 1) * (256 + MULTx) */
+        uint32_t m = osrval_p1 * 256;
+        for (uint32_t multx_p256 = 256;
+             multx_p256 < 256 + UART_MULTX_MAX && best_error != 0 &&
+             m / 2 <= target;
+             multx_p256++, m += osrval_p1) {
+            uint32_t error = (target + m / 2) % m;
+            error = abs((int32_t)error - (int32_t)(m / 2));
+            if (error < best_error) {
+                /* Only in this case we need to do the division as well. */
+                uint32_t brgval_p1 = (target + m / 2) / m;
+                if (brgval_p1 > (UART_BRGVAL_MAX + 1)) {
+                    continue;
+                }
+                best_osrval = osrval_p1 - 1;
+                best_multx = multx_p256 - 256;
+                best_brgval = brgval_p1 - 1;
+                best_error = error;
+            }
+        }
+        if (best_error == 0) {
+            break;
+        }
+    }
+
+    if (best_osrval == 0) {
+        return UART_NOBAUD;
+    }
+
+    CLOCK_SetClkDiv(flexcomm_num ? kCLOCK_DivFrg1 : kCLOCK_DivFrg0, best_multx);
+    dev->BRG = best_brgval;
+    dev->OSR = best_osrval;
+    return UART_OK;
+}
+
+int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
+{
+    assert(uart < UART_NUMOF);
+
+    const uart_conf_t *uart_conf = uart_config + uart;
+    USART_Type *const dev = uart_conf->dev;
+
+    int flexcomm_num = flexcomm_init((FLEXCOMM_Type *)dev, FLEXCOMM_ID_UART);
+    if (flexcomm_num < 0) {
+        return UART_INTERR;
+    }
+    uart_dev_from_flexcomm[flexcomm_num] = uart;
+
+    int ret = _uart_set_baudrate(dev, flexcomm_num, baudrate);
+    if (ret != UART_OK) {
+        return ret;
+    }
+
+    /* remember callback addresses */
+    config[uart].rx_cb = rx_cb;
+    config[uart].arg = arg;
+
+    /* Interrupt trigger and level for RX and TX disabled by default. */
+    dev->FIFOTRIG = 0;
+
+    /* Enable RX side. */
+    if (rx_cb != NULL && uart_conf->rx_pin != GPIO_UNDEF) {
+        /* Trigger RX interrupt when there is at least 1 byte (RXLVL = 0). */
+        dev->FIFOTRIG |= USART_FIFOTRIG_RXLVLENA_MASK | USART_FIFOTRIG_RXLVL(0);
+
+        /* Enable RX interrupt. */
+        dev->FIFOCFG |= USART_FIFOCFG_EMPTYRX_MASK |
+                        USART_FIFOCFG_ENABLERX_MASK;
+        dev->FIFOINTENSET = USART_FIFOINTENSET_RXLVL_MASK;
+        /* flexcomm_num is the same as the USART instance number in the
+         * USART_IRQS array. */
+        const uint8_t usart_irqn[] = USART_IRQS;
+        NVIC_EnableIRQ(usart_irqn[flexcomm_num]);
+    }
+    else {
+        dev->FIFOCFG &= ~USART_FIFOCFG_ENABLERX_MASK;
+        uart_conf->dev->FIFOINTENCLR = USART_FIFOINTENSET_RXLVL_MASK;
+    }
+
+    /* Enable TX side. */
+    if (uart_conf->tx_pin != GPIO_UNDEF) {
```

Comment:
```suggestion
    if (gpio_is_valid(uart_conf->tx_pin)) {
```

for future-proving this for the GPIO API update

Reply:
added "gpio.h" and changed these two to use `gpio_is_valid()`.